# Project : Linear Regression

Project Desrcription:
In this project I will find the best fitting line to a group of points. To minimize error, this line should be drawn as closely to each point as possible. 
Steps:
1. Function find_y(): Calculates the y value of a point that lies on a line. Given values: point's x value, line's slope and intercept.
2. Function error_y(): Finds the error which is the vertical distance of a point to a line. Given values: line's slope and intercept, point's x, y values as a tuple.
3. Function total_error(): Calculates the total error of all the points to the line. Given values: Line's slope and intercept, a list of points
4. Generate hundreds of lines with varying slopes and intercepts.
4. Finally, from the generated lines, find the line which gives the smallest error for a given set of datapoints

1. Find y value of a point that lies on a line:

In [1]:
def find_y(m, b, x):
    y = m * x + b
    return y

2. Find a point's distance to a line:

In [2]:
def error_y(m, b, point):
    #First, for the point's x value, find the line's corresponding y value.
    y_line = find_y(m, b, point[0])
    #Find the difference (error) between point's y value and the line's 
    #corresponding y value
    error = abs(point[1]-y_line)
    return error

Let's test the error_y() for the point (3, 5) and the line y = 2x + 4:

In [3]:
print(error_y(2, 4, (3, 5)))

5


3. Now we can find a point's distance to a line, but our goal is to find the total error given a list of points. Function total_error() will do that. 

In [4]:
def total_error(m, b, points):
    #Go over each point in points to find out it's distance to the line
    error_of_points = 0
    for point in points:
        error_of_points += error_y(m, b, point)
    return error_of_points
        

4. Function total_error() gives the total error for a **given** line and datapoints, but our goal eventually is to **find** a line that minimizes that error. To do that, we can generate hundreds of lines with varying slopes and intercepts, and find which combination returns the minimum error for a given groups of data. So, let's generate those lines with varying slopes and intercepts and also define a group of data points:

In [5]:
slopes = [ i * 0.01 for i in range(-200, 200)]
intercepts = [ i * 0.01 for i in range(-200, 200)]
point_group = [(1,3), (2,4), (4,1), (5,3), (4,4), (6,2)]

5. Now we will go through each slope and intercept and calculate the total distance (total error) to the points. We need to find the line (min_slope, min_intercept) that produces the smallest error (min_error):

In [6]:
min_slope = 0
min_intercept = 0
min_error = float("inf")
for slope in slopes:
    for intercept in intercepts:
        error_for_line = total_error(slope, intercept, point_group)
        if error_for_line < min_error:
            #Found a smaller error than the previous value
            #Record all values in case this is the smallest error
            min_error = error_for_line
            min_slope = slope
            min_intercept = intercept 
print("Slope: {:.2f}\nIntercept: {:.2f}\nMinimum Error: {:.2f} ".
      format(min_slope, min_intercept, min_error))

Slope: 0.20
Intercept: 1.99
Minimum Error: 6.62 


So the line y = 0.2x + 1.99 fits best to the points with an error of 6.62.

This project found the line that is closest to a given group of points.